src/App.js

```diff
@@ -18,27 +24,78 @@ const theme = createMuiTheme({
   }
 });
 
-export default function App() {
+const AuthRoute = props => {
+  if (!props.user.roles.includes('UNAUTHORIZED')) return <Redirect to="/" />;
+  return <Route {...props} />;
+};
 
+const App = () => {
   const [errors, setErrors] = useState({});
-  const [user, setUser] = useState({});
+  const [user, setUser] = useState({ roles: ['UNAUTHORIZED'] });
+  const [loginError, setLoginError] = useState('');
+  const [loading, setLoading] = useState(false);
+
+  const classes = useStyles();
+
+  useEffect(() => {
+    axios.defaults.headers.common = {'Authorization': localStorage.getItem('token')}
+    getUserProfile();
+  }, [])
+
+  function userLogin(email, password) {
+    setLoginError('');
+    axios
+       .post(`${process.env.REACT_APP_SERVER_URL}/api/login`, { email: email, password: password })
+       .then(res => {
+          localStorage.setItem('token', `Bearer ${res.data.accessToken}`);
+          axios.defaults.headers.common = {'Authorization': localStorage.getItem('token')}
+          getUserProfile();
+       })
+       .catch(err => {
+          err.response ?
+          setLoginError(`${err.response.data.error}: ${err.response.data.message}`) :
+          setLoginError('System error, please try again later.');
+       });
+   }
+
+  function getUserProfile() {
+    setLoading(true);
+    axios
+    .get(`${process.env.REACT_APP_SERVER_URL}/api/user/profile`)
+    .then(res => {
+      setUser(res.data);
+      setLoading(false);
+    })
+    .catch(err => {
+      setLoading(false);
+      // setLoginError('Error while getting profile information, please try again later.');
+    });
+  }
 
   return (
     <UserContext.Provider value={[user, setUser]}>
       <ErrorsContext.Provider value={[errors, setErrors]}>
         <MuiThemeProvider theme={theme}>
-          <BrowserRouter>
+          {
+            loading ?
+            <CircularProgress size={100} thickness={5} className={classes.spinner}/> :
+            <BrowserRouter>
             <Switch>
               <Route exact path='/' component={MainPage} />
-              <Route path='/login' component={LoginPage} />
+              <AuthRoute path='/login' user={user}>
+                <LoginPage userLogin={userLogin} loginError={loginError}/>
+              </AuthRoute>
```

Comment:
I think we could have AuthUserRoute and AuthManagerRoute and that would be sufficient for our needs. So just extending this a bit would suit all our needs, I think?